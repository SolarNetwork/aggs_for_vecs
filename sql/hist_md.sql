-- smallint hist_md
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements;
-- smallint hist_md none
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
-- smallint hist_md one null
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
-- smallint hist_md array of nulls
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
-- smallint hist_md one not-null
SELECT hist_md(smallints, '{1,3}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
-- smallint hist_md array of nulls and one other
SELECT hist_md(smallints, '{1,3}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);

-- int hist_md
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements;
-- int hist_md none
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
-- int hist_md one null
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
-- int hist_md array of nulls
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
-- int hist_md one not-null
SELECT hist_md(ints, '{1,3}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
-- int hist_md array of nulls and one other
SELECT hist_md(ints, '{1,3}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint hist_md
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements;
-- bigint hist_md none
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
-- bigint hist_md one null
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
-- bigint hist_md array of nulls
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
-- bigint hist_md one not-null
SELECT hist_md(bigints, '{1,3}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
-- bigint hist_md array of nulls and one other
SELECT hist_md(bigints, '{1,3}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);

-- real hist_md
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements;
-- real hist_md none
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
-- real hist_md one null
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
-- real hist_md array of nulls
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
-- real hist_md one not-null
SELECT hist_md(reals, '{1,3}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
-- real hist_md array of nulls and one other
SELECT hist_md(reals, '{1,3}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 hist_md
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements;
-- float8 hist_md none
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
-- float8 hist_md one null
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
-- float8 hist_md array of nulls
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
-- float8 hist_md one not-null
SELECT hist_md(floats, '{1,3}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
-- float8 hist_md array of nulls and one other
SELECT hist_md(floats, '{1,3}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);
