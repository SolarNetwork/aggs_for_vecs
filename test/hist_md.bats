load test_helper

@test "int16 hist_md" {
  result="$(query "SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements")";
  echo $result;
  [ "$result" = "{{0,0,0},{0,1,0},{0,2,0}}" ]
}

# @test "int32 hist_md" {
  # result="$(query "SELECT hist_md('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 hist_md" {
  # result="$(query "SELECT hist_md('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 hist_md" {
  # result="$(query "SELECT hist_md('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 hist_md" {
  result="$(query "SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements")";
  echo $result;
  [ "$result" = "{{0,0,0},{0,1,0},{0,2,0}}" ]
}

@test "float8 hist_md none" {
  result="$(query "SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 hist_md one null" {
  result="$(query "SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "{{0,0,0},{0,0,0},{0,0,0}}" ]
}

@test "float8 hist_md array of nulls" {
  result="$(query "SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{{0,0,0},{0,0,0},{0,0,0}}" ]
}

@test "float8 hist_md one not-null" {
  result="$(query "SELECT hist_md(floats, '{1,3}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{{0,0,0},{0,0,0},{0,1,0}}" ]
}

@test "float8 hist_md array of nulls and one other" {
  result="$(query "SELECT hist_md(floats, '{1,3}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{{0,0,0},{0,0,0},{0,1,0}}" ]
}

@test "string hist_md" {
  run query "SELECT hist_md(vals, '{1,2}'::int[], '{a,a}'::text[], '{b,b}'::text[], '{3,3}'::int[]) FROM (VALUES (ARRAY['a', 'b']), (ARRAY['b', 'c'])) t(vals)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  hist_md vals must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION" ]
}
