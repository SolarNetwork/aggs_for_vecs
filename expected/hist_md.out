-- smallint hist_md
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements;
          hist_md          
---------------------------
 {{0,0,0},{0,1,0},{0,2,0}}
(1 row)

-- smallint hist_md none
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
 hist_md 
---------
 
(1 row)

-- smallint hist_md one null
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- smallint hist_md array of nulls
SELECT hist_md(smallints, '{1,2}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- smallint hist_md one not-null
SELECT hist_md(smallints, '{1,3}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- smallint hist_md array of nulls and one other
SELECT hist_md(smallints, '{1,3}'::int[], '{0,0}'::smallint[], '{1,1}'::smallint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- int hist_md
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements;
          hist_md          
---------------------------
 {{0,0,0},{0,1,0},{0,2,0}}
(1 row)

-- int hist_md none
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
 hist_md 
---------
 
(1 row)

-- int hist_md one null
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- int hist_md array of nulls
SELECT hist_md(ints, '{1,2}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- int hist_md one not-null
SELECT hist_md(ints, '{1,3}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- int hist_md array of nulls and one other
SELECT hist_md(ints, '{1,3}'::int[], '{0,0}'::int[], '{1,1}'::int[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- bigint hist_md
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements;
          hist_md          
---------------------------
 {{0,0,0},{0,1,0},{0,2,0}}
(1 row)

-- bigint hist_md none
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
 hist_md 
---------
 
(1 row)

-- bigint hist_md one null
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- bigint hist_md array of nulls
SELECT hist_md(bigints, '{1,2}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- bigint hist_md one not-null
SELECT hist_md(bigints, '{1,3}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- bigint hist_md array of nulls and one other
SELECT hist_md(bigints, '{1,3}'::int[], '{0,0}'::bigint[], '{1,1}'::bigint[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- real hist_md
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements;
          hist_md          
---------------------------
 {{0,0,0},{0,1,0},{0,2,0}}
(1 row)

-- real hist_md none
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
 hist_md 
---------
 
(1 row)

-- real hist_md one null
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- real hist_md array of nulls
SELECT hist_md(reals, '{1,2}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- real hist_md one not-null
SELECT hist_md(reals, '{1,3}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- real hist_md array of nulls and one other
SELECT hist_md(reals, '{1,3}'::int[], '{0,0}'::real[], '{1,1}'::real[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- float8 hist_md
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements;
          hist_md          
---------------------------
 {{0,0,0},{0,1,0},{0,2,0}}
(1 row)

-- float8 hist_md none
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = -1;
 hist_md 
---------
 
(1 row)

-- float8 hist_md one null
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 1;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- float8 hist_md array of nulls
SELECT hist_md(floats, '{1,2}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 2;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,0,0}}
(1 row)

-- float8 hist_md one not-null
SELECT hist_md(floats, '{1,3}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id = 4;
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

-- float8 hist_md array of nulls and one other
SELECT hist_md(floats, '{1,3}'::int[], '{0,0}'::float[], '{1,1}'::float[], '{3,3}'::int[]) FROM measurements WHERE sensor_id IN (2, 4);
          hist_md          
---------------------------
 {{0,0,0},{0,0,0},{0,1,0}}
(1 row)

